crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,499 @@
+            // see earlier comments in this function
+            let first_item = self
+                .items
+                .first()
+                .expect("Expected there to be at least two items in the list");
+            let first_item_line_offset = locator.line_start(first_item.start());
+            if first_item_line_offset == locator.line_start(self.start()) {
+                first_item.start()
+            } else {
+                first_item_line_offset
+            }
+        };
+        let (needs_trailing_comma, postlude_start) = {
+            // We should already have returned by now if there are 0 items:
+            // see earlier comments in this function
+            let last_item = self
+                .items
+                .last()
+                .expect("Expected there to be at least two items in the list");
+            let last_item_line_offset = locator.line_end(last_item.end());
+            if last_item_line_offset == locator.line_end(self.end()) {
+                (false, last_item.end())
+            } else {
+                (true, last_item_line_offset)
+            }
+        };
+        let mut prelude = locator
+            .slice(TextRange::new(self.start(), prelude_end))
+            .to_string();
+        let postlude = locator.slice(TextRange::new(postlude_start, self.end()));
+
+        let joined_items = if self.multiline {
+            let indentation = stylist.indentation();
+            let newline = stylist.line_ending().as_str();
+            prelude = format!("{}{}", prelude.trim_end(), newline);
+            join_multiline_dunder_all_items(
+                &sorted_items,
+                locator,
+                parent,
+                indentation,
+                newline,
+                needs_trailing_comma,
+            )
+        } else {
+            Some(join_singleline_dunder_all_items(&sorted_items, locator))
+        };
+
+        let new_dunder_all = joined_items.map(|items| format!("{prelude}{items}{postlude}"));
+        SortedDunderAll::Sorted(new_dunder_all)
+    }
+}
+
+impl Ranged for DunderAllValue<'_> {
+    fn range(&self) -> TextRange {
+        *self.range
+    }
+}
+
+#[derive(Debug)]
+enum SortedDunderAll {
+    AlreadySorted,
+    Sorted(Option<String>),
+}
+
+fn collect_dunder_all_lines(range: TextRange, locator: &Locator) -> Option<Vec<DunderAllLine>> {
+    // Collect data on each line of `__all__`.
+    // Return `None` if `__all__` appears to be invalid,
+    // or if it's an edge case we don't care about.
+    let mut parentheses_open = false;
+    let mut lines = vec![];
+    let mut items_in_line = vec![];
+    let mut comment_in_line = None;
+    for pair in lexer::lex(locator.slice(range).trim(), Mode::Expression) {
+        let (tok, subrange) = pair.ok()?;
+        match tok {
+            Tok::Lpar | Tok::Lsqb => {
+                if parentheses_open {
+                    return None;
+                }
+                parentheses_open = true;
```

Comment:
Can you say more about the bracket handling here? I'm trying to follow. Typically, we do something more like this, where we track all open and closing brackets via a counter, to know if we're at the top level: https://github.com/astral-sh/ruff/blob/eb4ed2471b48d6ccba74064c65e37e32cba41a6c/crates/ruff_python_parser/src/soft_keywords.rs#L67. In this case, we're aborting if we see nested brackets, is that right?

Reply:
> In this case, we're aborting if we see nested brackets, is that right?

Yup, that's correct. If there are any parenthesized items in `__all__`, that would required a lot more complexity to handle, and it seems like a massive edge case, so not worth it.

> should `Lbrace` also be included?

It's currently handled by the catch-all `_` condition at the end of the `match` (which just returns `None` if we encounter any token not explicitly enumerated already), but I can move it here if you like?

Reply:
In that case, though, why  is `Lsqb` handled here?

Reply:
Because `__all__` could be a list, and we don't want to reject the opening square bracket of `__all__` if it's a list

Reply:
Added a big comment clarifying this in 572bb66546b9d0307b0de8e1e576041890410396 :)